libusb/os/windows_winusb.c

```diff
@@ -1450,6 +1450,140 @@ static int set_hid_interface(struct libusb_context *ctx, struct libusb_device *d
+
+	// the 1st pass tries to get the guid. if it fails due to ERROR_MORE_DATA
+	// reallocate enough memory for the 2nd pass
+	for(pass = 0; pass < 2; pass++) {
+		// Look for both DeviceInterfaceGUIDs *and* DeviceInterfaceGUID, in that order
+		// If multiple GUIDs, find the n-th that is indexed by guid_number
+		s = pRegQueryValueExA(key, "DeviceInterfaceGUIDs", NULL, &reg_type,
+			(LPBYTE)guid_string, &size);
+		if (s == ERROR_FILE_NOT_FOUND)
+			s = pRegQueryValueExA(key, "DeviceInterfaceGUID", NULL, &reg_type,
+				(LPBYTE)guid_string, &size);
+		if ( s == ERROR_SUCCESS) {
+			//the GUID was read successfully
+			break;
+		}
+		else if (s == ERROR_FILE_NOT_FOUND) {
+			usbi_warn(ctx, "no DeviceInterfaceGUID registered for '%s'", dev_id);
+			err = LIBUSB_ERROR_ACCESS;
+			goto exit;
+		} else if (s == ERROR_MORE_DATA) {
+			if (pass == 1) {
+				//there should have been allocated enough memory, but reading failed
+				usbi_warn(ctx, "unexpected error from pRegQueryValueExA for '%s'", dev_id);
+				err = LIBUSB_ERROR_OTHER;
+				goto exit;
+			}
+			// Allocate enough memory for storing the guid_string with two extra terminating characters
+			// it is necessary for parsing the REG_MULTI_SZ type bellow
+			new_guid_string = realloc((void *)guid_string, size + 2);
+			if (new_guid_string == NULL) {
+				usbi_err(ctx, "failed to realloc guid string");
+				err = LIBUSB_ERROR_NO_MEM;
+				goto exit;
+			}
+			guid_string = new_guid_string;
+		} else {
+			usbi_warn(ctx, "unexpected error from pRegQueryValueExA for '%s'", dev_id);
+			err = LIBUSB_ERROR_ACCESS;
+			goto exit;
+		}
+	}
+
+	// https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexa#remarks
+	// - "string may not have been stored with the proper terminating null characters"
+	// - The following GUIDs should be consider as valid:
+	// "{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}\0", "{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}",
+	// "{xxx.....xx}\0\0\0", "{xxx.....xx}\0{xxx.....xx}\0{xxx.....xx}\0",
+	// "{xxx.....xx}\0{xxx.....xx}\0{xxx.....xx}", "{xxx.....xx}{xxx.....xx}{xxx.....xx}",
+	// "{xxx.....xx}\0{xxx.....xx}\0{xxx.....xx}\0\0\0\0"
+	if ((reg_type == REG_SZ ) || (reg_type == REG_MULTI_SZ)) {
+		/* get the n-th guid indexed by guid_number since the DeviceInterfaceGUIDs may
+		   contain more GUIDs */
+		guid = guid_string;
+		// add two terminating chars for not overrunning the allocated memory during iterating the guid
+		guid[size] = '\0';
+		guid[size + 1] = '\0';
+		// iterating the guid
+		while(guid_number) {
+			guid = strchr(guid, '}');
+			if (guid == NULL) {
+				usbi_warn(ctx, "no GUID with index %d registered for '%s'", guid_number, dev_id);
+				err = LIBUSB_ERROR_ACCESS;
+				goto exit;
+			}
+			guid++;
+			// skip the terminating char if available
+			if (*guid == '\0') {
+				guid++;
+			}
+
+			guid_number--;
+		}
+		// add terminating char to the string
+		tmp_str = strchr(guid, '}');
+		if ((tmp_str == NULL) || (guid_number > 0)) {
```

Comment:
Isn't guid_number always zero at this point, after the while loop? So also the below warning will always say "index 0" ?

Reply:
Yes, you're right, the guid_number is always zero at this point. And thus the warning message prints a wrong guid number.

Reply:
Does my last commit sort this out correctly? I cannot test this.

Reply:
I performed a few tests with all supported variants. All my tests passed, the guid index is correctly printed now.

Reply:
Thanks for testing (again)!